File: problems/chapter13/mshibatatt/006.py
# find cycle in directed graph
# https://qiita.com/drken/items/a803d4fc4a727e02f7ba
import sys
sys.setrecursionlimit = 1000000

def dfs(G, n, seen, finished, path, pos):
    seen[n] = True
    path.append(n)
    for nxt in G[n]:
        if finished[nxt]:
            continue
        if seen[nxt] == True and finished[nxt] == False:
            pos = nxt
            return pos
        pos = dfs(G, nxt, seen, finished, path, pos)
        if pos != -1:
            return pos
    finished[n] = True
    path.pop()

def main():
    N, M = map(int, input().split())
    G = [[] for _ in range(N)]
    for _ in range(M):
        a, b = map(int, input().split())
        a -= 1
        b -= 1
        G[a].append(b)

    seen = [False]*N
    finished = [False]*N
    output = 'No'
    path = []
    pos = -1

    while not all(seen):
        for i, e in enumerate(seen):
            if not e:
                node = i
                break
        
        pos = dfs(G, node, seen, finished, path, pos)
        if pos != -1:
            output = 'Yes'
            cycle = []
            top = path[-1]
            while len(path) > 0:
                t = path[-1]
                cycle.append(t)
                path.pop()
                if t == pos:
                    cycle.append(top)
                    break
            break

    if output == 'Yes':
        print(output, ' -> '.join(map(str, cycle)))
    else:
        print(output)

if __name__=='__main__':
    main()

File: problems/chapter13/miiitomi/006.cpp
// https://qiita.com/drken/items/996d80bcae64649a6580#3-5-%E3%82%B5%E3%82%A4%E3%82%AF%E3%83%AB%E6%A4%9C%E5%87%BA
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=GRL_4_A&lang=jp

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

vector<bool> seen;
vector<bool> finished;
stack<int> S;
int pos = -1;

void dfs(const vector<vector<int>> &G, int v) {
    seen[v] = true;
    S.push(v);

    for (int nv : G[v]) {
        if (finished[nv]) continue;

        if (seen[nv]) {
            pos = nv;
            return;
        }

        dfs(G, nv);
        if (pos != -1) return;
    }
    S.pop();
    finished[v] = true;
}

int main() {
    int N, M;
    cin >> N >> M;
    vector<vector<int>> G(N);
    for (int i = 0; i < M; i++) {
        int a, b;
        cin >> a >> b;
        G[a].push_back(b);
    }

    seen.assign(N, false);
    finished.assign(N, false);

    for (int i = 0; i < N; i++) {
        if (!finished[i]) dfs(G, i);
        if (pos != -1) break;
    }

    if (pos == -1) cout << "No" << endl;
    else {
        vector<int> cycle;
        while (S.top() != pos) {
            cycle.push_back(S.top());
            S.pop();
        }
        cycle.push_back(pos);
        for (int i = (int)cycle.size() - 1; i >= 0; i--) cout << cycle[i] << " ";
        cout << endl;
    }
}
